rfcs/0019-storing-client-data.md

```diff
@@ -0,0 +1,484 @@
+
+```ocaml
+module Transition_frontier_diff = struct
+  type t =
+    | Breadcrumb_added of {block:  (block, State_hash.t) With_hash.t; sender_receipt_chains_from_parent_ledger: Receipt_chain_hash.t Public_key.Compressed.Map.t}
+    | Root_transitioned: of {new_: state_hash; garbage: state_hash.t list}
+    | Bootstrap of {lost_blocks : state_hash list}
+end
+```
+
+`Breadcrumb_added` has the field `sender_receipt_chains_from_parent_ledger` because we will use the previous `receipt_chain_hashes` from senders involved in sending a `user_command` in the new added breadcrumb to compute the `receipt_chain_hashes` of the transactions that appear in the breadcrumb. More of this will be discussed later.
+
+Whenever a node is bootstrapping, it will produce a diff containing the `state_hash` of all the blocks it has. These blocks would be lost and their consensus state in the database would be labeled as UNKNOWN.
+
+The `transaction_pool` diffs will essentially have a set of transactions that got added and removed from the `transaction_pool` after an operation has been done onto it
+
+```ocaml
+module Transaction_pool_diff = struct
+  type t = {added: User_command.t Set.t ; removed: User_command.t Set.t}
+end
+```
+
+Whenever the daemon produces a new `Transition_frontier_diff`, it will batch the transactions into the buffer, `transition_frontier_diff_pipe`. The buffer will have a max size. Once the size of the buffer exceeds the max size, it will flush the diffs to the archive process. To make sure that the archive process applies the diff with a low amount of latency, the buffer will flush the diffs in a short interval (like every 5 seconds). There will be a similar buffering process for the `transaction_pool`. The `transaction_pool` diffs can be seen as a monoid and it would be trivial to concate them into one diff. This would send less memory to the archive process and ultimately reduce the amount of writes that have to hit to the databases.
+
+It is worthy to note that whenever the archive process makes a write to the databases using deffereds. The archive process will write the diffs it receives in a sequential manner and will not context switch to another task. This will prevent any consistency issues.
```

Comment:
The first sentence is a little confusing. Should it be read as
```markdown
It is worth noting that the archive process writes to the databases using `Deferred.t`s.
```
or some other way?